src/gba/vfame.c

```diff
@@ -0,0 +1,303 @@
+		value = (addr & 0xFFFF) ^ 0x5555;
+		break;
+	case 0x090000:
+		value = ((addr & 0xFFFF) ^ 0x5555) - 1;
+		break;
+	case 0x0A0000:
+	case 0x0B0000:
+		value = _patternRightShift2(addr);
+		break;
+	case 0x0C0000:
+	case 0x0D0000:
+		value = 0xFFFF - _patternRightShift2(addr);
+		break;
+	case 0x0E0000:
+	case 0x0F0000:
+		value = _patternRightShift2(addr) ^ 0xAAAA;
+		break;
+	case 0x100000:
+	case 0x110000:
+		value = _patternRightShift2(addr) ^ 0x5555;
+		break;
+	case 0x120000:
+		value = 0xFFFF - ((addr & 0xFFFF) >> 1);
+		break;
+	case 0x130000:
+		value = 0xFFFF - ((addr & 0xFFFF) >> 1) - 0x8000;
+		break;
+	case 0x140000:
+	case 0x150000:
+		value = _patternRightShift1(addr) ^ 0xAAAA;
+		break;
+	case 0x160000:
+	case 0x170000:
+		value = _patternRightShift1(addr) ^ 0x5555;
+		break;
+	case 0x180000:
+	case 0x190000:
+		value = _patternRightShift1(addr) ^ 0xF0F0;
+		break;
+	case 0x1A0000:
+	case 0x1B0000:
+		value = _patternRightShift1(addr) ^ 0x0F0F;
+		break;
+	case 0x1C0000:
+	case 0x1D0000:
+		value = _patternRightShift1(addr) ^ 0xFF00;
+		break;
+	case 0x1E0000:
+	case 0x1F0000:
+		value = _patternRightShift1(addr) ^ 0x00FF;
+		break;
+	}
+
+	return value & 0xFFFF;
+}
+
+static uint32_t _patternRightShift1(uint32_t addr) {
+	return (addr >> 1) & 0xFFFF;
+}
+
+static uint32_t _patternRightShift2(uint32_t addr) {
+	uint32_t value = addr & 0xFFFF;
+	value = value >> 2;
+	value += (addr & 3) == 2 ? 0x8000 : 0;
+	value += (addr & 0x10000) ? 0x4000 : 0;
+	return value;
+}
+
+void GBAVFameSramWrite(struct GBAVFameCart* cart, uint32_t address, int8_t value, uint8_t* sramData) {
+	// A certain sequence of writes to SRAM FFF8->FFFC can enable or disable "mode change" mode
+	// Currently unknown if these writes have to be sequential, or what happens if you write different values, if anything
+	if (address >= 0x0E00FFF8 && address <= 0xE00FFFC) {
+		cart->writeSequence[address - 0x0E00FFF8] = value;
+		if (address == 0xE00FFFC) {
+			if (memcmp(MODE_CHANGE_START_SEQUENCE, cart->writeSequence, sizeof(MODE_CHANGE_START_SEQUENCE)) == 0){
+				cart->acceptingModeChange = true;
+			}
+			if (memcmp(MODE_CHANGE_END_SEQUENCE, cart->writeSequence, sizeof(MODE_CHANGE_END_SEQUENCE)) == 0){
+				cart->acceptingModeChange = false;
+			}
+		}
+	}
+
+	// If we are in "mode change mode" we can change either SRAM or ROM modes
+	// Currently unknown if other SRAM writes in this mode should have any effect
+	if (cart->acceptingModeChange) {
+		if (address == 0x0E00FFFE) {
+			cart->sramMode = value;
+		} else if (address == 0xE00FFFD) {
+			cart->romMode = value;
+		}
+	}
+
+	if (cart->sramMode == -1) {
+		// when SRAM mode is uninitialised you can't write to it
+		return;
+	}
+
+	// if mode has been set - the address and value of the SRAM write will be modified
+	address = _modifySramAddress(cart->cartType, address, cart->sramMode);
+	value = _modifySramValue(cart->cartType, value, cart->sramMode);
+	// these writes are mirrored
+	if (address >= 0xE008000) {
```

Comment:
`address &= 0x7FFF` works without needing a check, and removes the need to do the masking below.